### PR TITLE
ENH: add figure.legend; outside kwarg for better layout outside subplots

### LINKED ISSUE #13023: constrained_layout support for figure.legend
Just a feature request to have constrained_layout support `figure.legend`

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

If constrained_layout is being used, this new kwarg allows the user to automagically place a legend outside the subplots in a grid spec (or all the subplots in a  figure for a more straightforward example).

Closes #13023  ping @dcherian

### simple example

```python
fig, axs = plt.subplots(1, 2, constrained_layout=True)

for i, ax in enumerate(axs):
    ax.plot(range(10), label=f'Boo{i}')
lg = fig.legend(loc='upper right', outside=True)
plt.show()
```

yields:

simple

### Nested gridspec example 

```python
fig = plt.figure(constrained_layout=True)
gs0 = fig.add_gridspec(1, 2)

gs = gs0[0].subgridspec(1, 1)
for i in range(1):
    ax = fig.add_subplot(gs[i,0])
    ax.plot(range(10), label=f'Boo{i}')
lg = fig.legend(ax=[ax], loc='upper right', outside=True, borderaxespad=4)

gs2 = gs0[1].subgridspec(3, 1)
axx = []
for i in range(3):
    ax = fig.add_subplot(gs2[i, 0])
    ax.plot(range(10), label=f'Who{i}', color=f'C{i+1}')
    if i < 2:
        ax.set_xticklabels('')
    axx += [ax]
lg2 = fig.legend(ax=axx[:-1], loc='upper right', outside=True, borderaxespad=4)
plt.show()
```

yields:

complicated

### Added ability to vertically lay-out the corners

See `outside='vertical'` below...

```python 
import matplotlib.pyplot as plt

fig = plt.figure(constrained_layout=True)
gs0 = fig.add_gridspec(1, 2)

gs = gs0[0].subgridspec(1, 1)
for i in range(1):
    ax = fig.add_subplot(gs[i,0])
    ax.plot(range(10), label=f'Boo{i}')
lg = fig.legend(ax=[ax], loc='upper right', outside=True, borderaxespad=4)

gs2 = gs0[1].subgridspec(3, 1)
axx = []
for i in range(3):
    ax = fig.add_subplot(gs2[i, 0])
    ax.plot(range(10), label=f'Who{i}', color=f'C{i+1}')
    if i < 2:
        ax.set_xticklabels('')
    axx += [ax]
lg2 = fig.legend(ax=axx[:-1], loc='upper right', outside='vertical',
                 borderaxespad=4, ncol=2)
plt.show()
```

yields:

figure_1

- [ ] needs more tests

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by timhoffm** [comment]
I‘m positive on the idea. But don‘t have the time to look into this in more detail for a few days.

So just a quick API thought for now: Is it reasonable to have two separate functions fig.legend and fig.legend_outside for the user? Internally, this is quite different, but it might be worth considering having something like fig.legend(loc=„outside upper right“) and branch internally. That of course depends on how similar the API of both versions would be.

(Sorry for the bad formatting - phone keyboard)

**Comment by jklymak** [comment]
> I‘m positive on the idea. But don‘t have the time to look into this in more detail for a few days.
> 
> So just a quick API thought for now: Is it reasonable to have two separate functions fig.legend and fig.legend_outside for the user? Internally, this is quite different, but it might be worth considering having something like fig.legend(loc="outside upper right") and branch internally. That of course depends on how similar the API of both versions would be.

The API can be identical from my point of view, except for specifying "outside" somehow.  Its not currently, but thats just because I think we are trying to get away from implicit APIs, and parsing the arguments so that the inputs work at different levels is a bit of a pain, but it can be dealt with. 

So, 

1. `fig.legend_outside(loc='upper right')`
2. `fig.legend(loc='outside upper right')`, but what about numerical codes `fig.legend(loc=1)`?
3. `fig.legend(loc='upper right', outside=True)` and `fig.legend(loc=1, outside=True)` with an error raised if `constrained_layout == False` or `bbox_to_anchor` is not `None`.

I'd be fine w/ 3, but I'm not sure what is more discoverable for users, 1 or 3.   Other opinions?

**Comment by dcherian** [comment]
I like 3. I'd be more likely to discover it by looking at the docstring for figure.legend

**Comment by pharshalp** [comment]
+1.0 for option 3, 0.0 for option 2,  -1.0 for option 1.

**Comment by anntzer** [comment]
+1 for option 3.

**Comment by jklymak** [comment]
... option 3 it is.

**Comment by anntzer** [comment]
Actually looking at it again... "upper right" can mean one of two things: in the upper-right corner, to the right of the axes, or in the upper-right corner, above the axis.
Looks like right now it's always to the right?  Should perhaps introduce new locations "right-upper" etc. so that "upper-right" means "UR corner + above"?  (Perhaps would read better using compass locations (#12679), then one can have "eastnortheast"/"ENE" vs "northnortheast"/"NNE"...)
A quick look suggests that MATLAB only has "northeastoutside" which behaves like "upper-right" does right now.

**Comment by jklymak** [comment]
Well since we now have the outside kwarg perhaps it could take a string value outside=“vertical”?

**Comment by anntzer** [comment]
How would it combine with "east"(/"left")?
Would be a bit ugly to have "outside" be True/False for "90°" directions and "vertical"/"horizontal" for the 45° ones.
Sorry for the nitpicking...

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -1727,6 +1726,18 @@ def legend(self, *args, **kwargs):
...
+        outside: bool or string
+            If ``constrained_layout=True``, then try and place legend outside
+            axes listed in *axs*, or highest-level gridspec if axs is empty.
+            Note, "center" and "best" options to *loc* do not work with
+            ``outside=True``.  The corner values of *loc* (i.e. "upper right")
+            will default to a horizontal layout of the legend, but this can
+            be changed by specifying a string
+            ``outside="vertical", loc="upper right"``.
+
+        axs :  sequence of `~.axes.Axes`
```
Comment (dcherian) [inline lib/matplotlib/figure.py]: does this need to be `axs` instead of `axes`?

Comment (anntzer) [reply lib/matplotlib/figure.py]: axs would be standard I think (axes is not great as it suggests a single Axes...)

Comment (jklymak) [reply lib/matplotlib/figure.py]: I guess `colorbar` uses `ax`.  Since this is supposed to be in parallel to that maybe I should change to that?

Comment (anntzer) [reply lib/matplotlib/figure.py]: sure, sounds fair

Comment (dcherian) [reply lib/matplotlib/figure.py]: ya that sounds good.

**Comment by jklymak** [comment]
Links to docs (as of previous push; noticed a couple of mistakes)

https://17015-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure

https://17015-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/text_labels_and_annotations/figlegendoutside_demo.html#sphx-glr-gallery-text-labels-and-annotations-figlegendoutside-demo-py

**Comment by anntzer** [comment]
Haven't read the other threads, but you may also want to check https://github.com/matplotlib/matplotlib/issues/3745 https://github.com/matplotlib/matplotlib/issues/3857 https://github.com/matplotlib/matplotlib/pull/6182.

**Comment by jklymak** [comment]
@anntzer,  thanks - those are largely for axes legends, whereas this is for figure legends.  OTOH, it would be easy to add an "outside" kwarg for axes legends as well, in fact easier than for figures, I think so long as constrained_layout is being used.  

*Both* of them can be done w/o constained_layout as well doing the same things that `colorbar` does, but that could be the next step of this.

**Comment by jklymak** [comment]
This latest push lets axes.legend have an outside=True as well.     It *doesn't* put the legend outside any axes decorations, so it can overlap tick labels, titles, etc.  This is because constrained_layout doesn't do anything below the axes level, and this is an axes level-artist.  But it is somewhat nicer than specifying the location by hand (?).


```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(constrained_layout=True)
ax.plot(range(10), label='boo')
ax.legend(loc='NE', outside=True)
plt.show()
```

axfigure_1

File: lib/matplotlib/gridspec.py [hunk 2]
```diff
@@ -170,6 +172,68 @@ def _normalize(key, size):  # Includes last index.
 
         return SubplotSpec(self, num1, num2)
 
+    def legend_outside(self, handles=None, labels=None, axs=None,
+                       align='horizontal', **kwargs):
+        """
+        legend for this gridspec, offset from all the subplots.
+
+        See `.Figure.legend_outside` for details on how to call.
```
Comment (anntzer) [inline lib/matplotlib/gridspec.py]: "the *outside* argument to `Figure.legend`"
Dunno whether you want to rename that function now that Figure.legend_outside was renamed, but perhaps not.

Comment (jklymak) [reply lib/matplotlib/gridspec.py]: Ooops, no this definitely still needs some documentation work.  Also the above new stuff is a bit of a mess still, so let me mark WIP.

**Comment by timhoffm** [comment]
Sorry for jumping in again at a late stage of the discussion, but I feel this needs thorough consideration to end up with a reasonable API.

#### Parameters

> Actually looking at it again... "upper right" can mean one of two things: in the upper-right corner, to the right of the axes, or in the upper-right corner, above the axis.
Looks like right now it's always to the right? Should perhaps introduce new locations "right-upper" etc. so that "upper-right" means "UR corner + above"? (Perhaps would read better using compass locations (#12679), then one can have "eastnortheast"/"ENE" vs "northnortheast"/"NNE"...)
A quick look suggests that MATLAB only has "northeastoutside" which behaves like "upper-right" does right now.

> Well since we now have the outside kwarg perhaps it could take a string value outside=“vertical”?

> How would it combine with "east"(/"left")?
Would be a bit ugly to have "outside" be True/False for "90°" directions and "vertical"/"horizontal"/False for the 45° ones.

This all tells me that two arguments are arguments are maybe not a great idea. We just have one task: *Defining the position of the legend.* Splitting this in a direction (like "upper right" or "northeast") and in inside/outside is artificial and even not clear because there is no unambiguous "upper right" for "outside". Also the following combinations do not have a defined meaning:

~~~
loc='best', outside=True
loc='center', outside=True
loc=(0.1, 0.1), outside=True
~~~

Also, we already have the meaning of `loc` depend on `bbox_to_anchor`. It's not getting simpler to document and understand if we have multiple interdependent parameters.

I have to think about the desired API a bit more.

#### Alignment

Looking at the first figure in https://17015-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/text_labels_and_annotations/figlegendoutside_demo.html, I feel that we actually would need some sort of alignment with the axes but outside:

- The left legend should be centered with the ylabel.
- The top legend should be centered between the two axes patches.
- The bottom of the right legend should be aligned with the bottom of the axes patch, not with the bottom of the ticks.

I currently don't have a clear strategy how to do this, but it is at least something to consider when defining the API.

File: lib/matplotlib/gridspec.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/gridspec.py]: Fixed....

**Comment by jklymak** [comment]
Are there further comments on the API?  Again, my tendency would be to have a different function `figure.legend_outside` and then we don't have kwarg clashes with `figure.legend`.  There is no way I can or want to make this work with `bbox_to_anchor`, and obviously `best`, `center` etc won't have any meaning.  Could have those cases error out.

**Comment by anntzer** [comment]
I think the fact that the supported kwargs are effectively different in the inside and outside cases makes me change my vote to be in favor of `legend_outside`.

And then you can actually have a new list of locations that has ENE/NNE (upperupperleft/leftupperleft), and not NE (upperleft) to handle the corners properly without having to guess...

**Comment by jklymak** [comment]
Ok ENE means upper right beside?  NNE means upper right above?   Did we want NE ( above and beside?).

**Comment by jklymak** [comment]
(@anntzer BTW this failed w the font bug: https://travis-ci.org/matplotlib/matplotlib/jobs/486672509)

**Comment by anntzer** [comment]
Thanks for the pointer, opened https://github.com/matplotlib/matplotlib/pull/13331.

**Comment by jklymak** [comment]
ping @timhoffm @dcherian for comments too.

**Comment by anntzer** [comment]
upper right vs right upper is cute.

**Comment by jklymak** [comment]
...Except 95% of people are going to *want* what you are calling “outside right upper” but specify it the more normal “outside upper right” and be sad when they don’t get that.

**Comment by timhoffm** [comment]
The idea behind this is that everything starting with "right" is essentially to the right and the second word is just a smaller modification of the vertical position. Is that too sophisticated?

I would also say, that most people would want "outside right". That's the default in Excel/LibreOffice (not that this is necessarily the best, but at least it's common).

**Comment by jklymak** [comment]
I understand.  Its just not idiomatic English to say "right upper corner" and that is the more common use case here.

**Comment by jklymak** [comment]
Still not sure what to do  about this one.  Seemed that consensus was for a separate method, but @timhoffman felt strongly that it should be a parsable part of the location kwarg of the main legend function. That’s somewhat held up by the fact that we aren’t entirely happy with our location kwarg strings.  

I am genuinely curious how often fig.legend is ever called, and if anyone ever wants it to *not* be outside the axes.  I imagine most real world examples are anchoring it outside at axes and then using bbox_inches=tight or adjusting the subplots so the legend fits.  

I guess so far this is a decent proof of concept. But I’m not sure if I like the fragility if it all.  If we were to add this functionality to non constrained layout should we just do the anchoring I suspect most people want?  Getting this to work without constrained_layout is maybe a good way to start reinvigorating it.

**Comment by ImportanceOfBeingErnest** [comment]
*This is just to answer the question on frequency of usage (I haven't thought about how this relates to or impacts this PR)*

> I am genuinely curious how often `fig.legend` is ever called, and if anyone ever wants it to *not* be outside the axes.

A direct use case is adding a single legend for multiple axes. E.g. in [Secondary axis with twinx(): how to add to legend?](https://stackoverflow.com/a/47370214/4124317) I proposed to use `fig.legend()` to get a single legend for twinned axes. Especially,

     fig.legend(loc="upper right", bbox_to_anchor=(1,1), bbox_transform=ax.transAxes)

to get that legend inside the axes. That answer has become quite popular by now. 

Also, the most popular solution to [How do I make a single legend for many subplots with matplotlib?](https://stackoverflow.com/questions/9834452/how-do-i-make-a-single-legend-for-many-subplots-with-matplotlib) mentions `fig.legend()` (though it wouldn't be clear from that answer  if people put it in- or outside any of the axes).

**Comment by timhoffm** [comment]
@jklymak wrong user mentioned (https://github.com/matplotlib/matplotlib/pull/13072#issuecomment-515804405 :smile:). I could go with a separate method `legend_outside` for now. However, the separate method does still not help with the location string.

Do we have a way of introducing experimental/preliminary API? If so, I'd suggest to do so here since we're not quite clear how the API would be best designed.

**Comment by timhoffm** [comment]
Semi-OT random thoughts:
- What "outside" legends really need to do is steal space from one or more axes. We have a similar concept in [colorbar](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.figure.Figure.html?highlight=colorbar#matplotlib.figure.Figure.colorbar) where the parameter `ax` can take one or more axes to steal space from. Maybe the concepts there and here are similar enough to have a common approach.
- Maybe we should have a non-Axes container that can be placed into a gridspec, which then holds the legend (or the legend could be placed inside its own axes - which OTOH sounds like a bad idea). Main point here would be that the legend gets a defined share of the layout.

**Comment by jjnurminen** [comment]
FWIW, I'm doing grids of subplots with duplicated curves (i.e. a curve may appear in one or more subplots). In this case, using a figure-level legend seems natural, as the legend is not associated with any particular plot. Also, `constrained_layout` works very well for me, except it doesn't support `fig.legend`. Thus, I've resorted to creating an extra 'legend axis' outside my subplots and manually constructing a legend there, but that's a bit clumsy. So I would be +1 on implementing this.

**Comment by jklymak** [comment]
Closed for #19743

**Pull request closed without merging**